## Instructions

1) Please watch recitation videos by TAs and my lecture video on Jupyternotebook before doing this lab.

2) After finishing the lab, save files as html and ipynb. 

3) To submit HW solutions, go to our course in Canvas --> Assignments --> HW 1 --> Submit Assignment --> Click "Chose file" --> choose HTML file you saved --> Click another file--> choose ipynb file--submit.

4) Once you submit, download,  and check to make sure that you have submitted the correct files. 

5) When workingfor the solutions to this lab, please do not display unnecessary data unless it is asked.  

6) Please do not share codes or the solutions in the piazza. We can still ask and clarify questions.

7) Private messages to your Instructors and TAs should be used to ask questions about codes and other details.

8) Each problem is worth 10 points.


#### Each problem is worth 10 points.

In [1]:
#import packages
import pandas as pd
import numpy as np

### Problem 1. 
Upload the .csv file titanic.csv that came with this HW in the notebook, and name it as titanic. Make sure to import panda as pd and numpy as np.  Those are the necessary packages. Read the details about the titanic data here https://data.world/nrippner/titanic-disaster-dataset. Show the head of the data and find out how many rows and columns are in the data.

In [2]:
titanic = pd.read_csv('titanic.csv', decimal =',')
titanic.head()

pclass  survived                                             name     sex  \
0     1.0       1.0                    Allen, Miss. Elisabeth Walton  female   
1     1.0       1.0                   Allison, Master. Hudson Trevor    male   
2     1.0       0.0                     Allison, Miss. Helen Loraine  female   
3     1.0       0.0             Allison, Mr. Hudson Joshua Creighton    male   
4     1.0       0.0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

      age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0      29    0.0    0.0   24160  211.3375       B5        S    2    NaN   
1  0.9167    1.0    2.0  113781  151.5500  C22 C26        S   11    NaN   
2       2    1.0    2.0  113781  151.5500  C22 C26        S  NaN    NaN   
3      30    1.0    2.0  113781  151.5500  C22 C26        S  NaN  135.0   
4      25    1.0    2.0  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON

In [3]:
titanic.shape

(1310, 14)

### Problem 2. 
 Make a list of variables in the titanic data set. Then classify all the variables as one of discrete, continuous, nominal, and ordinal. Explain how you determined the types of variables.

In [4]:
list(titanic.columns)

['pclass',
 'survived',
 'name',
 'sex',
 'age',
 'sibsp',
 'parch',
 'ticket',
 'fare',
 'cabin',
 'embarked',
 'boat',
 'body',
 'home.dest']

In [5]:
titanic.pclass = titanic.pclass.astype(object)
titanic.survived = titanic.survived.astype(bool)
titanic.age = titanic.age.astype(float)
titanic.fare = titanic.fare.astype(float)

pclass - ordinal, survived - nominal, name - nominal, sex - nominal, age - continuous, sibsp - discrete, parch - discrete, ticket - nominal, fare - continuous, cabin - ordinal, embarked - nominal, boat - nominal, body - nominal, home dest - nominal. If they are qualitative, but can be ordered, then they are ordinal. If they are qualitative and can't be ordered, then they are nominal. If they are quantitative and continuous, then they are continuous. If they are quantitative and not continuous, then they are discrete.

### Problem 3. 
 Learn more about isnull() or isna() function  here https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.isnull.html and use it to find how many missing values does each variable has. Find out which three variables have the most missing values, remove them from the data frame, and name the new data frame as titanic1.

In [6]:
titanic.isna().sum()

pclass          1
survived        0
name            1
sex             1
age           264
sibsp           1
parch           1
ticket          1
fare            2
cabin        1015
embarked        3
boat          824
body         1189
home.dest     565
dtype: int64

Body, Cabin, and Boat have the most na values

In [7]:
titanic1 = titanic.dropna(axis ='columns',thresh = 700)

### Problem 4. 
Filter the titanic1 data with and operator. https://kanoki.org/2020/01/21/pandas-dataframe-filter-with-multiple-conditions/

a) How many passengers who paid less than the average fare survived? 

b) Did a passenger who embarked journey from Queenstown and was of 40 years of age survive?


In [8]:
titanic1.loc[(titanic1['fare']<np.mean(titanic1.fare)) & (titanic1['survived'] == True)]

pclass  survived                                             name  \
5         1      True                              Anderson, Mr. Harry   
14        1      True             Barkworth, Mr. Algernon Henry Wilson   
22        1      True                            Behr, Mr. Karl Howell   
29        1      True        Bjornstrom-Steffansson, Mr. Mauritz Hakan   
31        1      True                                 Blank, Mr. Henry   
...     ...       ...                                              ...   
1261      3      True                           Turkula, Mrs. (Hedwig)   
1277      3      True                             Vartanian, Mr. David   
1286      3      True  Whabee, Mrs. George Joseph (Shawneene Abi-Saab)   
1290      3      True                 Wilkes, Mrs. James (Ellen Needs)   
1300      3      True          Yasbeck, Mrs. Antoni (Selini Alexander)   

         sex   age  sibsp  parch  ticket     fare embarked  \
5       male  48.0    0.0    0.0   19952  26.5500        S   
14      male  80.0    0.0    0.0   27042  30.0000        S   
22      male  26.0    0.0    0.0  111369  30.0000        C   
29      male  28.0    0.0    0.0  110564  26.5500        S   
31      male  40.0    0.0    0.0  112277  31.0000        C   
...      ...   ...    ...    ...     ...      ...      ...   
1261  female  63.0    0.0    0.0    4134   9.5875        S   
1277    male  22.0    0.0    0.0    2658   7.2250        C   
1286  female  38.0    0.0    0.0    2688   7.2292        C   
1290  female  47.0    1.0    0.0  363272   7.0000        S   
1300  female  15.0    1.0    0.0    2659  14.4542        C   

                               home.dest  
5                           New York, NY  
14                         Hessle, Yorks  
22                          New York, NY  
29    Stockholm, Sweden / Washington, DC  
31                        Glen Ridge, NJ  
...                                  ...  
1261                                 NaN  
1277                                 NaN  
1286                                 NaN  
1290                                 NaN  
1300                                 NaN  

[319 rows x 11 columns]

In [9]:
titanic1.loc[(titanic1['age'] == 40) & (titanic1['home.dest'].str.startswith('Queenstown')) & (titanic1['survived'] == True)]

Empty DataFrame
Columns: [pclass, survived, name, sex, age, sibsp, parch, ticket, fare, embarked, home.dest]
Index: []

a)319
b)No

### Problem 5. 
Continuing with the titanic1 data set, answer the following questions.

a) Use groupby function to find out how many passengers were in each class. https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html

b) Create a new column with a column name “luck”  in the titanic1 data set with the following rules. Let’s call a passenger “unlucky” if the passenger was in the first-class, paid more than the average fare, and did not survive.  On the contrary, a passenger is “lucky” if the passenger was not in the first class, paid less or equal to the average fare, and survived. Let’s call everybody else, “averageluck.”  Show the head of the data with a visible luck column.


In [10]:
titanic1.groupby('pclass').size()

pclass
1.0    323
2.0    277
3.0    709
dtype: int64

In [11]:
titanic1['luck'] = 'averageluck'

<ipython-input-11-683827a69197>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  titanic1['luck'] = 'averageluck'


In [12]:
titanic1[(titanic1.fare > np.mean(titanic1.fare)) & (titanic1.pclass == 1) & (titanic1.survived == False)].replace(['averageluck'],'unlucky')

pclass  survived                                             name  \
2       1.0     False                     Allison, Miss. Helen Loraine   
3       1.0     False             Allison, Mr. Hudson Joshua Creighton   
4       1.0     False  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
9       1.0     False                          Artagaveytia, Mr. Ramon   
10      1.0     False                           Astor, Col. John Jacob   
..      ...       ...                                              ...   
310     1.0     False                         Wick, Mr. George Dennick   
312     1.0     False                       Widener, Mr. George Dunton   
313     1.0     False                        Widener, Mr. Harry Elkins   
316     1.0     False                      Williams, Mr. Charles Duane   
318     1.0     False           Williams-Lambert, Mr. Fletcher Fellows   

        sex   age  sibsp  parch    ticket      fare embarked  \
2    female   2.0    1.0    2.0    113781  151.5500        S   
3      male  30.0    1.0    2.0    113781  151.5500        S   
4    female  25.0    1.0    2.0    113781  151.5500        S   
9      male  71.0    0.0    0.0  PC 17609   49.5042        C   
10     male  47.0    1.0    0.0  PC 17757  227.5250        C   
..      ...   ...    ...    ...       ...       ...      ...   
310    male  57.0    1.0    1.0     36928  164.8667        S   
312    male  50.0    1.0    1.0    113503  211.5000        C   
313    male  27.0    0.0    2.0    113503  211.5000        C   
316    male  51.0    0.0    1.0  PC 17597   61.3792        C   
318    male   NaN    0.0    0.0    113510   35.0000        S   

                            home.dest     luck  
2     Montreal, PQ / Chesterville, ON  unlucky  
3     Montreal, PQ / Chesterville, ON  unlucky  
4     Montreal, PQ / Chesterville, ON  unlucky  
9                 Montevideo, Uruguay  unlucky  
10                       New York, NY  unlucky  
..                                ...      ...  
310                    Youngstown, OH  unlucky  
312                   Elkins Park, PA  unlucky  
313                   Elkins Park, PA  unlucky  
316  Geneva, Switzerland / Radnor, PA  unlucky  
318                   London, England  unlucky  

[77 rows x 12 columns]

In [13]:
titanic1[(titanic1.fare < np.mean(titanic1.fare)) & (titanic1.pclass != 1) & (titanic1.survived == True)].replace(['averageluck'],'lucky')

pclass  survived                                               name  \
324      2.0      True              Abelson, Mrs. Samuel (Hannah Wizosky)   
329      2.0      True  Angle, Mrs. William A (Florence "Mary" Agnes H...   
333      2.0      True                            Ball, Mrs. (Ada E Hall)   
336      2.0      True                                  Beane, Mr. Edward   
337      2.0      True                  Beane, Mrs. Edward (Ethel Clarke)   
...      ...       ...                                                ...   
1261     3.0      True                             Turkula, Mrs. (Hedwig)   
1277     3.0      True                               Vartanian, Mr. David   
1286     3.0      True    Whabee, Mrs. George Joseph (Shawneene Abi-Saab)   
1290     3.0      True                   Wilkes, Mrs. James (Ellen Needs)   
1300     3.0      True            Yasbeck, Mrs. Antoni (Selini Alexander)   

         sex   age  sibsp  parch     ticket     fare embarked  \
324   female  28.0    1.0    0.0  P/PP 3381  24.0000        C   
329   female  36.0    1.0    0.0     226875  26.0000        S   
333   female  36.0    0.0    0.0      28551  13.0000        S   
336     male  32.0    1.0    0.0       2908  26.0000        S   
337   female  19.0    1.0    0.0       2908  26.0000        S   
...      ...   ...    ...    ...        ...      ...      ...   
1261  female  63.0    0.0    0.0       4134   9.5875        S   
1277    male  22.0    0.0    0.0       2658   7.2250        C   
1286  female  38.0    0.0    0.0       2688   7.2292        C   
1290  female  47.0    1.0    0.0     363272   7.0000        S   
1300  female  15.0    1.0    0.0       2659  14.4542        C   

                             home.dest   luck  
324                Russia New York, NY  lucky  
329                   Warwick, England  lucky  
333   Bristol, Avon / Jacksonville, FL  lucky  
336             Norwich / New York, NY  lucky  
337             Norwich / New York, NY  lucky  
...                                ...    ...  
1261                               NaN  lucky  
1277                               NaN  lucky  
1286                               NaN  lucky  
1290                               NaN  lucky  
1300                               NaN  lucky  

[277 rows x 12 columns]

In [14]:
titanic1.head()

pclass  survived                                             name     sex  \
0      1      True                    Allen, Miss. Elisabeth Walton  female   
1      1      True                   Allison, Master. Hudson Trevor    male   
2      1     False                     Allison, Miss. Helen Loraine  female   
3      1     False             Allison, Mr. Hudson Joshua Creighton    male   
4      1     False  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

       age  sibsp  parch  ticket      fare embarked  \
0  29.0000    0.0    0.0   24160  211.3375        S   
1   0.9167    1.0    2.0  113781  151.5500        S   
2   2.0000    1.0    2.0  113781  151.5500        S   
3  30.0000    1.0    2.0  113781  151.5500        S   
4  25.0000    1.0    2.0  113781  151.5500        S   

                         home.dest         luck  
0                     St Louis, MO  averageluck  
1  Montreal, PQ / Chesterville, ON  averageluck  
2  Montreal, PQ / Chesterville, ON  averageluck  
3  Montreal, PQ / Chesterville, ON  averageluck  
4  Montreal, PQ / Chesterville, ON  averageluck